src/fermilib/utils/_graph.py

```diff
@@ -0,0 +1,198 @@
+
+        del self.uid_to_index[node_uid]
+        del self.nodes[node_id]
+        del self.neighbors[node_id]
+        del self.node_uids[node_id]
+
+    def node_count(self):
+        """Number of nodes in the graph
+
+        Returns(int): Number of nodes currently in the graph
+        """
+        return len(self.nodes)
+
+    def add_edge(self, node_id1, node_id2):
+        """Add an edge between node 1 and node2
+
+        Args:
+            node_id1(int): Index of first node on edge
+            node_id2(int): Index of second node on edge
+        """
+        if node_id1 >= self.node_count() or node_id2 >= self.node_count():
+            raise IndexError("Node ID out of range for graph")
+
+        self.neighbors[node_id1].add(self.node_uids[node_id2])
+        self.neighbors[node_id2].add(self.node_uids[node_id1])
+
+    def remove_edge(self, node_id1, node_id2):
+        """Remove an edge between node1 and node2
+
+        Args:
+            node_id1(int): Index of first node on edge
+            node_id2(int): Index of second node on edge
+        """
+        if node_id1 >= self.node_count() or node_id2 >= self.node_count():
+            raise IndexError("Node ID out of range for graph")
+
+        self.neighbors[node_id1].remove(self.node_uids[node_id2])
+        self.neighbors[node_id2].remove(self.node_uids[node_id1])
+
+    def is_adjacent(self, node_id1, node_id2):
+        """Test for adjacency between node1 and node2
+
+         Args:
+            node_id1(int): Index of first node
+            node_id2(int): Index of second node
+        """
+        if node_id1 >= self.node_count() or node_id2 >= self.node_count():
+            raise IndexError("Node ID out of range for graph")
+
+        return (self.node_uids[node_id1] in self.neighbors[node_id2])
+
+    def get_neighbors(self, node_id):
+        """Return list of neighbors of the specified node
+
+        Args:
+            node_id: Index of node to examine the neighbors of
+
+        Returns(list): List of current node IDs that are neighbors of node_id.
+        """
+        if node_id >= self.node_count():
+            raise IndexError("Node ID out of range for graph")
+        return [self.uid_to_index[i] for i in self.neighbors[node_id]]
+
+    def shortest_path(self, node_id1, node_id2):
+        """Find the shortest path between node1 and node2 on the graph
+
+        Args:
+            node_id1(int): Index of first node
+            node_id2(int): Index of second node
+
+        Returns(list): List of nodes from node_id1 to node_id2 that constitute
+        the shortest possible path in the graph between those two nodes.
+        """
+        if node_id1 >= self.node_count() or node_id2 >= self.node_count():
+            raise IndexError("Node ID out of range for graph")
+
+        # Treat special case of equal inputs
+        if node_id1 == node_id2:
+            return [node_id1]
+
+        previous = [None] * self.node_count()
```

Comment:
Can you please add a couple of comments here to explain what is going on with code in this function? I understand it now but I had to look at it for a few minutes and one or two comments would have been very helpful.